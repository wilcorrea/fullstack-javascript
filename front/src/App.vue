<template>
  <div id="app">
    <!-- where the magic happens -->
    <!-- @see front/src/router/index.ts:22 -->
    <router-view />
  </div>
</template>

<script lang="js">
import $store from 'src/store'
import Local from 'src/app/DataSource/Local'

/**
 */
export default {
  /**
   */
  name: 'App',
  /**
   */
  data: () => ({
    // state to map the version update confirmation
    confirmed: false
  }),
  /**
   */
  computed: {
    /**
     * Get the version from store
     * @see createStore
     * @returns {string}
     */
    version () {
      return $store.state.version
    }
  },
  /**
   */
  methods: {
    /**
     * This is a simple method to ask user to update the version
     * @param {string} message
     */
    notify (message) {
      if (this.confirmed) {
        return
      }
      this.confirmed = window.confirm(message)
      if (this.confirmed) {
        window.location.reload(true)
      }
    },
    /**
     * Receive the version fetched from server
     * @param {AxiosResponse} result
     */
    receiveVersion (result) {
      if (this.version && this.version !== result.data) {
        const message = `We have updates!` +
          ` The current version is ${this.version} and the new one is ${result.data}.` +
          ` Do you wanna reload the app now?`
        this.notify(message)
      }
      $store.commit('updateVersion', result.data)
    },
    /**
     * Fetch the current app version from server
     */
    fetchVersion () {
      Local
        .instance()
        .get('version')
        .then(this.receiveVersion)
    }
  },
  /**
   */
  created () {
    // wait 1 second to take a breath
    window.setTimeout(this.fetchVersion, 1000)
  }
}
</script>
