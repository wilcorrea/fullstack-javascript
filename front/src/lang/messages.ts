import ptBr from 'src/lang/pt-br'
import { Record } from 'src/app/Agnostic/interfaces'

/**
 * The standard messages list
 * @type {Record}
 */
export const messages: Record = { 'pt-br': ptBr }

/**
 * Set the messages
 * Can be used to replace or add new messages
 * @param {string} alias
 * @param {Record} newLocale
 */
export const setMessages = (alias: string, newLocale: Record) => {
  messages[alias] = newLocale
}

/**
 * Get the messages
 * @param {string} alias
 * @returns {Record}
 */
export const getMessages = (alias: string): Record => {
  return messages[alias]
}

/**
 * Export getMessages as default
 * @param {string} locale
 * @returns {Record}
 */
export default getMessages
