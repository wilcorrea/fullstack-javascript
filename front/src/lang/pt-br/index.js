import app from './app'
import domains from 'src/domains/pt-br'
import pages from './pages'
import views from './views'

/**
 * put all messages together
 */
export default { app, domains, pages, views }
