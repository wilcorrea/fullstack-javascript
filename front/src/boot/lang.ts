import $lang from 'src/lang'

/**
 * @param {Vue} Vue
 * @returns {}
 */
export default (Vue: any) => {
  /**
   */
  Object.defineProperty(Vue.prototype, '$lang', {
    get () {
      return $lang
    }
  })
}
