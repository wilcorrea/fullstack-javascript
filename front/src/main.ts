import Vue from 'vue'

import router from 'src/router'
import store from 'src/store/vuex'
import Boot from 'src/boot'
// import 'src/registerServiceWorker'

import 'material-icons/iconfont/material-icons.css'
import 'src/assets/loading.css'
import 'src/assets/themes/default/@index.styl'

import App from 'src/App.vue'
import 'src/app/Components'

Vue.config.productionTip = false

Vue.use(Boot)

new Vue({
  router,
  store,
  render: h => h(App)
}).$mount('#app')
