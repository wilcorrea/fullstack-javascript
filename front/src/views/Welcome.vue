<template>
  <div class="app-layout__page Welcome">
    <!-- use i18n to get the title -->
    <h1>{{ $lang('views.welcome.title') }}</h1>
    <div v-html="$lang('views.welcome.content')" />
    <br>
    <!-- print the version that is generated at each deploy -->
    <pre>version {{ version }}</pre>
  </div>
</template>

<script lang="js">
import $store from 'src/store'

/**
 * @component {Welcome}
 */
export default {
  /**
   */
  name: 'Welcome',
  /**
   * Wrapper to version that is in a store
   */
  computed: {
    /**
     * @returns {string}
     */
    version () {
      return $store.state.version
    }
  }
}
</script>
