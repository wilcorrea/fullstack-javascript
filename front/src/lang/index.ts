import messages from './messages'
import { get } from 'src/app/Util/general'
import { i18n } from 'src/settings'

/**
 * Start with a default locale
 * @type {string}
 */
let locale = i18n()

/**
 * Set the current locale
 * @param {string} newLocale
 */
export const setLocale = (newLocale: string) => {
  locale = newLocale
}

/**
 * Get the current locale
 * @returns {string}
 */
export const getLocale = (): string => {
  return locale
}

/**
 * Lang is a featured i18n engine
 * We can change the locale and add more messages in realtime
 * I do not like some things in vue-i18n and other libs that I've tested, so that's it
 * @param {string} path
 * @param {string} fallback
 * @returns {*}
 */
export default (path: string, fallback: string = ''): any => {
  return get(messages(locale), path, fallback)
}
