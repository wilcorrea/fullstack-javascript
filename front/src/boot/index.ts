import lang from './lang'
import browse from './browse'
import util from './util'
import './polyfill'

/**
 * @type {{install: Function}}
 */
export default {
  install (Vue: any) {
    lang(Vue)
    browse(Vue)
    util(Vue)
  }
}
