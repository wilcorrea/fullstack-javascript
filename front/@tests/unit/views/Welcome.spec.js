import { shallowMount } from '@vue/test-utils'
import Welcome from 'src/views/Welcome.vue'
import $lang from 'src/lang'

describe('Welcome', () => {
  const wrapper = shallowMount(Welcome, { mocks: { $lang } })

  /**
   * @test
   */
  it('check component markup', () => {
    expect(wrapper.html()).toMatch(`<h1>${$lang('views.welcome.title')}</h1>`)
  })

  /**
   * @test
   */
  test('should render content correctly', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })
})
