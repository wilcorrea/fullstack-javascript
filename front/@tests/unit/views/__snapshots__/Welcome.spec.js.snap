// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Welcome should render content correctly 1`] = `
<div class="app-layout__page Welcome">
  <h1>Seja bem vindo!</h1>
  <div>
    <p></p>
    <h3>Olá, tudo bem por ai? Espero que sim</h3>
    <p></p>
    <p>Você está tendo acesso à este projeto que é um desafio que fiz para uma avaliação
      para uma posição de Tech Lead Front na Wirecard Brasil, então vou explicar umas coisinhas
      sobre isso.</p>
    <p>
      Foi um grande prazer ter feito o teste e confesso que foi até divertido (em alguns momentos).
      Lidar com tarefas com escopo próximo do real sem compromisso é uma experiência que
      raramente temos.. estamos sempre com pressa para alguma coisa e acaba que a gente não
      para pra ver como as coisas podem ser legais :/
    </p>
    <p>
      Sem mais delongas, vamos ao que interessa! Vou tentar explicar um pouco para você
      (avaliador ou apenas alguém está dando uma olhadinha) como o projeto se passa.
    </p>
    <p>
      O repositório tem duas pastas, como se fossem dois projetos mesmo, que
      funcionam de forma independente: <b><i>back</i></b> e <b><i>front</i></b>.
      Na pasta back você poderá encontrar um serviço criado com express que consome
      uma api REST e entrega o conteúdo através de GraphQL. Já no front temos um projeto
      criado com o Vue-CLI com base bem simples. Nos dois foram aplicadas o mínimo
      de dependências que eu consegui (mas sempre dá para otimizar né... aceito dicas =)
      e sempre que deu tentei usar várias abordagens diferentes para uma mesma coisa.
    </p>
    <p>
      Este projeto é antes de tudo uma demonstração da forma com a qual eu trabalho
      com VueJS, Javascript, TypeScript, Stylus, e toda essa sopa de letrinhas.
      Provavelmente o front estará mais bonito que o back, mas a a oportunidade em
      questão era pra o front, então foi onde centrei meus esforços.
    </p>
    <br>
    <p>
      Na <a href="/#/dashboard/about">página Sobre</a> eu explico melhor sobre as
      tecnologias aplicadas. Dá uma chegada lá para conferior. Você pode clicar neste
      link aqui detrás ou usar o "menuzinho" maroto que está pulsando lá em cima.
    </p>
    <br><br>
    <small>
      P.S.: este número estranho aqui embaixo é a versão do app. Ele muda toda vez
      que rola um deploy, também falo dele lá na <a href="/#/dashboard/about">página Sobre</a>!
    </small>
  </div> <br> <pre>version </pre>
</div>
`;
